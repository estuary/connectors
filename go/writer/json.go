package writer

import (
	"compress/flate"
	"fmt"
	"io"

	"github.com/estuary/connectors/go/encrow"
	"github.com/klauspost/pgzip"
	"github.com/segmentio/encoding/json"
)

const (
	// JSON generally compresses well at minimum compression levels. Higher levels of compression
	// will usually take a lot more CPU while not providing much space savings.
	jsonCompressionlevel = flate.BestSpeed

	// Snowflake docs (https://docs.snowflake.com/en/sql-reference/sql/put#usage-notes) recommend
	// data file sizes be in the range of 100-250MB for compressed data. Redshift docs
	// (https://docs.aws.amazon.com/redshift/latest/dg/t_splitting-data-files.html) suggest files
	// for compressed data be in the range of 1MB to 1GB, and all be about the same size. Bigquery
	// docs don't mention anything other than the files must be less than 4GB. So, a 250MB file size
	// seems reasonable to use across the board given the current materializations that use this.
	DefaultJsonFileSizeLimit = 250 * 1024 * 1024
)

type jsonConfig struct {
	disableCompression bool
	skipNulls          bool
}

type JsonOption func(*jsonConfig)

func WithJsonDisableCompression() JsonOption {
	return func(cfg *jsonConfig) {
		cfg.disableCompression = true
	}
}

func WithJsonSkipNulls() JsonOption {
	return func(cfg *jsonConfig) {
		cfg.skipNulls = true
	}
}

type JsonWriter struct {
	w     io.Writer // will be set to `gz` for compressed writes or `cwc` if compression is disabled
	cwc   *countingWriteCloser
	gz    *pgzip.Writer
	shape *encrow.Shape
	buf   []byte
}

// NewJsonWriter creates a JsonWriter from w. w is closed when JsonWriter is closed. If `fields`
// is nil, values will be encoded as a JSON array rather than as an object.
func NewJsonWriter(w io.WriteCloser, fields []string, opts ...JsonOption) *JsonWriter {
	var cfg jsonConfig
	for _, o := range opts {
		o(&cfg)
	}

	jw := &JsonWriter{
		cwc: &countingWriteCloser{w: w},
	}

	if !cfg.disableCompression {
		gz, err := pgzip.NewWriterLevel(jw.cwc, jsonCompressionlevel)
		if err != nil {
			// Only possible if compressionLevel is not valid.
			panic("invalid compression level for gzip.NewWriterLevel")
		}
		jw.gz = gz
		jw.w = gz
	} else {
		jw.w = jw.cwc
	}

	if fields != nil {
		jw.shape = encrow.NewShape(fields)
		// Setting TrustRawMessage here prevents unnecessary validation of pre-serialized JSON
		// received from the runtime, which we can assume to be valid (flow_document for example).
		// Note that we are also not setting SortMapKeys or EscapeHTML: Sorting keys is not needed
		// because encrow.Shape already sorts the top-level keys and any object values are already
		// serialized as JSON, and escaping HTML is not desired so as to avoid escaping values like
		// <, >, &, etc. if they are present in the materialized collection's data.
		jw.shape.SetFlags(json.TrustRawMessage)
		jw.shape.SetSkipNulls(cfg.skipNulls)
	}

	return jw
}

func (w *JsonWriter) Write(vals []any) (err error) {
	w.buf = w.buf[:0]
	if w.shape == nil {
		// Serialize as a JSON array of values.
		w.buf = append(w.buf, '[')
		for idx, v := range vals {
			if w.buf, err = json.Append(w.buf, v, json.TrustRawMessage); err != nil {
				return fmt.Errorf("encoding JSON array value: %w", err)
			}
			if idx != len(vals)-1 {
				w.buf = append(w.buf, ',')
			}
		}
		w.buf = append(w.buf, ']')
	} else {
		// Serialize as a JSON object.
		if w.buf, err = w.shape.Encode(w.buf, vals); err != nil {
			return fmt.Errorf("encoding shape: %w", err)
		}
	}

	w.buf = append(w.buf, '\n')

	if _, err := w.w.Write(w.buf); err != nil {
		return fmt.Errorf("writing gzip bytes: %w", err)
	}

	return nil
}

func (w *JsonWriter) Written() int {
	return w.cwc.written
}

// Close closes the underlying gzip writer if compression is enabled, flushing its data and writing
// the GZIP footer. It also closes the underlying io.WriteCloser that was used to initialize the
// counting writer.
func (w *JsonWriter) Close() error {
	if w.gz != nil {
		if err := w.gz.Close(); err != nil {
			return fmt.Errorf("closing gzip writer: %w", err)
		}
	}

	if err := w.cwc.Close(); err != nil {
		return fmt.Errorf("closing counting writer: %w", err)
	}
	return nil
}
