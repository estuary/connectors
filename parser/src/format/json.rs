//! Parser for the json format. This will accept any stream of JSON values separated by whitespace.
//! It allows any amount of whitespace (including newlines) within and in between records.
use super::{Input, Output, ParseError, Parser};
use crate::ParseConfig;

struct JsonParser;

pub fn new_parser() -> Box<dyn Parser> {
    Box::new(JsonParser)
}

impl Parser for JsonParser {
    fn parse(&self, _config: &ParseConfig, content: Input) -> Result<Output, ParseError> {
        let deser = serde_json::de::Deserializer::from_reader(content).into_iter();
        let wrapped = deser.map(|r| r.map_err(Into::into));
        Ok(Box::new(wrapped))
    }
}
