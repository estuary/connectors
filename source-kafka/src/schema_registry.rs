use anyhow::Result;
use futures::stream::{self, StreamExt};
use reqwest::Client;
use serde::Deserialize;
use std::collections::{HashMap, HashSet};

const TOPIC_KEY_SUFFIX: &str = "-key";
const TOPIC_VALUE_SUFFIX: &str = "-value";
const CONCURRENT_SCHEMA_REQUESTS: usize = 10;

#[derive(Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
struct FetchedSchema {
    #[serde(default = "SchemaType::default")]
    schema_type: SchemaType,
    schema: String,
    reference: Option<SchemaReference>,
}

#[derive(Deserialize, Debug)]
struct SchemaReference {
    name: String,
    subject: String,
    version: u32,
}

#[derive(Deserialize, Debug)]
struct FetchedLatestVersion {
    id: u32,
}

#[derive(Deserialize, Debug)]
#[serde(rename_all = "UPPERCASE")]
enum SchemaType {
    Avro,
    Json,
    Protobuf,
}

impl SchemaType {
    fn default() -> Self {
        SchemaType::Avro
    }
}

#[derive(Debug)]
pub enum RegisteredSchema {
    Avro(apache_avro::Schema),
    Json(serde_json::Value),
    Protobuf, // TODO(whb): Protobuf support.
}

#[derive(Debug, Default)]
pub struct TopicSchema {
    pub key: Option<RegisteredSchema>,
    pub value: Option<RegisteredSchema>,
}

pub struct SchemaRegistryClient {
    endpoint: String,
    http: Client,
    username: String,
    password: String,
}

impl SchemaRegistryClient {
    pub fn new(endpoint: String, username: String, password: String) -> SchemaRegistryClient {
        SchemaRegistryClient {
            endpoint: endpoint.to_string(),
            http: reqwest::Client::default(),
            username,
            password,
        }
    }

    pub async fn schemas_for_topics(
        &self,
        topics: &[String],
    ) -> Result<HashMap<String, TopicSchema>> {
        let applicable_topics: HashSet<String> = topics.iter().cloned().collect();

        let subjects: Vec<String> = self
            .http
            .get(format!("{}/subjects", self.endpoint))
            .basic_auth(&self.username, Some(&self.password))
            .send()
            .await?
            .json()
            .await?;

        let filter_by_suffix = |s: &str, suffix: &str| {
            if let Some(s) = s.strip_suffix(suffix) {
                if !applicable_topics.contains(s) {
                    return None;
                }
                return Some(s.to_string());
            }
            None
        };

        let topics_with_key_schema: HashSet<String> = subjects
            .iter()
            .filter_map(|s| filter_by_suffix(s, TOPIC_KEY_SUFFIX))
            .collect();

        let topics_with_value_schema: HashSet<String> = subjects
            .iter()
            .filter_map(|s| filter_by_suffix(s, TOPIC_VALUE_SUFFIX))
            .collect();

        let schema_futures: Vec<_> = applicable_topics
            .iter()
            .filter_map(|topic| {
                let need_key = topics_with_key_schema.contains(topic);
                let need_value = topics_with_value_schema.contains(topic);
                if !need_key && !need_value {
                    return None;
                }
                Some(async move {
                    let mut schema = TopicSchema {
                        key: None,
                        value: None,
                    };

                    if need_key {
                        schema.key = Some(self.fetch_latest_schema(topic, true).await?)
                    }
                    if need_value {
                        schema.value = Some(self.fetch_latest_schema(topic, false).await?)
                    }

                    Ok::<(String, TopicSchema), anyhow::Error>((topic.to_owned(), schema))
                })
            })
            .collect();

        stream::iter(schema_futures)
            .buffer_unordered(CONCURRENT_SCHEMA_REQUESTS)
            .collect::<Vec<_>>()
            .await
            .into_iter()
            .collect::<Result<HashMap<String, TopicSchema>>>()
    }

    pub async fn fetch_schema(&self, id: u32) -> Result<RegisteredSchema> {
        let fetched: FetchedSchema = self
            .http
            .get(format!("{}/schemas/ids/{}", self.endpoint, id))
            .basic_auth(&self.username, Some(&self.password))
            .send()
            .await?
            .json()
            .await?;

        match fetched.schema_type {
            SchemaType::Avro => {
                // TODO(whb): Resolve references.
                let schema = apache_avro::Schema::parse_str(&fetched.schema)
                    .expect("failed to parse avro schema");
                Ok(RegisteredSchema::Avro(schema))
            }
            SchemaType::Json => {
                // TODO(whb): Resolve references for JSON too.
                let schema =
                    serde_json::from_str(&fetched.schema).expect("failed to parse json schema");
                Ok(RegisteredSchema::Json(schema))
            }
            SchemaType::Protobuf => Ok(RegisteredSchema::Protobuf),
        }
    }

    async fn fetch_latest_version(&self, subject: &str) -> Result<u32> {
        let fetched: FetchedLatestVersion = self
            .http
            .get(format!(
                "{}/subjects/{}/versions/latest",
                self.endpoint, subject
            ))
            .basic_auth(&self.username, Some(&self.password))
            .send()
            .await?
            .json()
            .await?;
        Ok(fetched.id)
    }

    async fn fetch_latest_schema(&self, topic: &str, key: bool) -> Result<RegisteredSchema> {
        let subject = format!(
            "{}{}",
            topic,
            if key {
                TOPIC_KEY_SUFFIX
            } else {
                TOPIC_VALUE_SUFFIX
            }
        );
        let version = self.fetch_latest_version(subject.as_str()).await?;
        self.fetch_schema(version).await
    }
}
