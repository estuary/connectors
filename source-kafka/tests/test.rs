use std::time::Duration;

use anyhow::Result;
use apache_avro::types::{Record, Value};
use apache_avro::Schema;
use rdkafka::admin::{AdminClient, AdminOptions, NewTopic, TopicReplication};
use rdkafka::message::{Header, OwnedHeaders, ToBytes};
use rdkafka::producer::{FutureProducer, FutureRecord};
use rdkafka::ClientConfig;
use schema_registry_converter::async_impl::avro::AvroEncoder;
use schema_registry_converter::async_impl::json::JsonEncoder;
use schema_registry_converter::async_impl::schema_registry::SrSettings;
use schema_registry_converter::schema_registry_common::SubjectNameStrategy;
use serde_json::json;

#[test]
fn test_spec() {
    let output = std::process::Command::new("flowctl")
        .args(["raw", "spec", "--source", "tests/test.flow.yaml"])
        .output()
        .unwrap();

    assert!(output.status.success());
    let got: serde_json::Value = serde_json::from_slice(&output.stdout).unwrap();
    insta::assert_snapshot!(serde_json::to_string_pretty(&got).unwrap());
}

#[tokio::test]
async fn test_discover() {
    setup_test().await;

    let output = std::process::Command::new("flowctl")
        .args([
            "raw",
            "discover",
            "--source",
            "tests/test.flow.yaml",
            "-o",
            "json",
            "--emit-raw",
        ])
        .output()
        .unwrap();

    assert!(output.status.success());

    let snap = std::str::from_utf8(&output.stdout)
        .unwrap()
        .lines()
        .map(|line| serde_json::from_str::<serde_json::Value>(line).unwrap())
        .map(|line| serde_json::to_string_pretty(&line).unwrap())
        .reduce(|snap, line| format!("{}\n{}", snap, line))
        .unwrap();

    insta::assert_snapshot!(snap);
}

#[tokio::test]
async fn test_capture() {
    setup_test().await;

    let output = std::process::Command::new("flowctl")
        .args([
            "preview",
            "--source",
            "tests/test.flow.yaml",
            "--sessions",
            "1",
            "--delay",
            "2s",
            "--output-state",
        ])
        .output()
        .unwrap();

    println!("{}", std::str::from_utf8(&output.stderr).unwrap());

    assert!(output.status.success());

    let snap = std::str::from_utf8(&output.stdout).unwrap();

    insta::assert_snapshot!(snap);
}

#[tokio::test]
async fn test_capture_resume() {
    setup_test().await;

    let initial_state = json!({
      "bindingStateV1": {
        "avro-topic": {
          "partitions": {
            "1": 1,
            "2": 1
          }
        },
        "json-schema-topic": {
          "partitions": {
            "1": 1,
            "2": 1
          }
        },
        "json-raw-topic": {
          "partitions": {
            "1": 1,
            "2": 1
          }
        }
      }
    });

    let output = std::process::Command::new("flowctl")
        .args([
            "preview",
            "--source",
            "tests/test.flow.yaml",
            "--sessions",
            "1",
            "--delay",
            "2s",
            "--output-state",
            "--initial-state",
            &initial_state.to_string(),
        ])
        .output()
        .unwrap();

    assert!(output.status.success());

    let snap = std::str::from_utf8(&output.stdout).unwrap();

    insta::assert_snapshot!(snap);
}

async fn setup_test() {
    let bootstrap_servers = "localhost:9092";
    let schema_registry_endpoint = "http://localhost:8081";
    let num_messages = 9;
    let num_partitions = 3;
    let topic_replication = 1;

    let test_cases: &[(&dyn TestDataEncoder, &str)] = &[
        (&AvroTestDataEncoder::new(), "avro-topic"),
        (&JsonSchemaTestDataEncoder::new(), "json-schema-topic"),
        (&JsonRawTestDataEncoder::new(), "json-raw-topic"),
    ];

    let http = reqwest::Client::default();

    let admin: AdminClient<_> = ClientConfig::new()
        .set("bootstrap.servers", bootstrap_servers)
        .create()
        .unwrap();

    let producer: FutureProducer = ClientConfig::new()
        .set("bootstrap.servers", bootstrap_servers)
        .create()
        .unwrap();

    let opts = AdminOptions::default().request_timeout(Some(Duration::from_secs(1)));

    for (enc, topic) in test_cases {
        admin.delete_topics(&[topic], &opts).await.unwrap();
        admin
            .create_topics(
                &[NewTopic::new(
                    topic,
                    num_partitions,
                    TopicReplication::Fixed(topic_replication),
                )],
                &opts,
            )
            .await
            .unwrap()
            .into_iter()
            .collect::<Result<Vec<_>, _>>()
            .unwrap();

        // Registry schemas if the encoder uses schemas.
        if let Some(schema_type) = enc.schema_type_string() {
            for (topic, suffix, schema) in [
                (topic, "key", &enc.key_schema_string()),
                (topic, "value", &enc.payload_schema_string()),
            ] {
                // Try to delete the existing schema if it exists, ignoring any errors
                // that are returned, which may be 404's.
                http.delete(format!(
                    "{}/subjects/{}-{}",
                    schema_registry_endpoint, topic, suffix
                ))
                .send()
                .await
                .unwrap();

                // You have to do a "soft" delete before a permanent "hard" delete.
                http.delete(format!(
                    "{}/subjects/{}-{}?permanent=true",
                    schema_registry_endpoint, topic, suffix
                ))
                .send()
                .await
                .unwrap();

                // Register the schema, which must be successful.
                assert!(http
                    .post(format!(
                        "{}/subjects/{}-{}/versions",
                        schema_registry_endpoint, topic, suffix
                    ))
                    .json(&json!({"schema": schema, "schemaType": schema_type}))
                    .send()
                    .await
                    .unwrap()
                    .status()
                    .is_success());
            }
        }

        // Populate regular "data" records.
        for idx in 0..num_messages {
            send_message(
                topic,
                &enc.key_for_idx(idx, topic).await,
                Some(&enc.payload_for_idx(idx, topic).await),
                idx,
                num_partitions,
                &producer,
            )
            .await;
        }

        // Populate deletion records.
        for idx in 0..num_partitions {
            send_message(
                topic,
                &enc.key_for_idx(idx as usize, topic).await,
                None::<&[u8]>,
                idx as usize,
                num_partitions,
                &producer,
            )
            .await;
        }
    }
}

async fn send_message<K, P>(
    topic: &str,
    key: &K,
    payload: Option<&P>,
    idx: usize,
    num_partitions: i32,
    producer: &FutureProducer,
) where
    K: ToBytes + ?Sized,
    P: ToBytes + ?Sized,
{
    let mut rec = FutureRecord::to(topic)
        .partition(idx as i32 % num_partitions)
        .key(key)
        .timestamp(unix_millis_fixture(idx))
        .headers(OwnedHeaders::new().insert(Header {
            key: "header-key",
            value: Some(&format!("header-value-{}", idx)),
        }));

    if let Some(payload) = payload {
        rec = rec.payload(payload);
    }

    producer.send(rec, None).await.unwrap();
}

fn unix_millis_fixture(idx: usize) -> i64 {
    ((idx + 1) * 86_400_000) as i64
}

#[async_trait::async_trait]
trait TestDataEncoder {
    async fn key_for_idx<'a>(&'a self, idx: usize, topic: &'a str) -> Vec<u8>;
    async fn payload_for_idx<'a>(&'a self, idx: usize, topic: &'a str) -> Vec<u8>;
    fn key_schema_string(&self) -> String;
    fn payload_schema_string(&self) -> String;
    fn schema_type_string(&self) -> Option<String>;
}

struct AvroTestDataEncoder {}

impl AvroTestDataEncoder {
    fn new() -> Self {
        AvroTestDataEncoder {}
    }
}

#[async_trait::async_trait]
impl TestDataEncoder for AvroTestDataEncoder {
    async fn key_for_idx<'a>(&'a self, idx: usize, topic: &'a str) -> Vec<u8> {
        let enc = AvroEncoder::new(SrSettings::new(String::from("http://localhost:8081")));
        let schema =
            Schema::parse(&serde_json::from_str(&self.key_schema_string()).unwrap()).unwrap();

        let mut key = Record::new(&schema).unwrap();
        key.put("idx", Value::Int(idx as i32));
        key.put(
            "nested",
            Value::Record(vec![("sub_id".to_string(), Value::Int(idx as i32))]),
        );

        enc.encode_value(
            key.into(),
            &SubjectNameStrategy::TopicNameStrategy(topic.to_string(), true),
        )
        .await
        .unwrap()
    }

    async fn payload_for_idx<'a>(&'a self, idx: usize, topic: &'a str) -> Vec<u8> {
        let enc = AvroEncoder::new(SrSettings::new(String::from("http://localhost:8081")));
        let schema =
            Schema::parse(&serde_json::from_str(&self.payload_schema_string()).unwrap()).unwrap();

        let mut value = Record::new(&schema).unwrap();
        value.put("value", Value::String(format!("value-{}", idx)));

        enc.encode_value(
            value.into(),
            &SubjectNameStrategy::TopicNameStrategy(topic.to_string(), false),
        )
        .await
        .unwrap()
    }

    fn key_schema_string(&self) -> String {
        let parsed = Schema::parse(&json!({
          "type": "record",
          "name": "AvroKey",
          "fields": [
            {
              "name": "idx",
              "type": "int"
            },
            {
              "name": "nested",
              "type": {
                "type": "record",
                "name": "NestedAvroKeyRecord",
                "fields": [
                  {
                    "name": "sub_id",
                    "type": "int"
                  }
                ]
              }
            }
          ]
        }))
        .unwrap();

        parsed.canonical_form()
    }

    fn payload_schema_string(&self) -> String {
        let parsed = Schema::parse(&json!({
          "type": "record",
          "name": "AvroValue",
          "fields": [
            {
              "name": "value",
              "type": "string"
            }
          ]
        }))
        .unwrap();

        parsed.canonical_form()
    }

    fn schema_type_string(&self) -> Option<String> {
        Some("AVRO".to_string())
    }
}

struct JsonSchemaTestDataEncoder {}

impl JsonSchemaTestDataEncoder {
    fn new() -> Self {
        Self {}
    }
}

#[async_trait::async_trait]
impl TestDataEncoder for JsonSchemaTestDataEncoder {
    async fn key_for_idx<'a>(&'a self, idx: usize, topic: &'a str) -> Vec<u8> {
        let enc = JsonEncoder::new(SrSettings::new(String::from("http://localhost:8081")));
        enc.encode(
            &json!({
                "idx": idx,
                "nested": {
                    "sub_id": idx
                },
            }),
            SubjectNameStrategy::TopicNameStrategy(topic.to_string(), true),
        )
        .await
        .unwrap()
    }

    async fn payload_for_idx<'a>(&'a self, idx: usize, topic: &'a str) -> Vec<u8> {
        let enc = JsonEncoder::new(SrSettings::new(String::from("http://localhost:8081")));
        enc.encode(
            &json!({
                "value": format!("value-{}", idx),
            }),
            SubjectNameStrategy::TopicNameStrategy(topic.to_string(), false),
        )
        .await
        .unwrap()
    }

    fn key_schema_string(&self) -> String {
        serde_json::to_string(&json!({
            "$schema": "http://json-schema.org/draft-07/schema#",
            "title": "JsonKey",
            "type": "object",
            "properties": {
              "idx": {
                "type": "integer"
              },
              "nested": {
                "type": "object",
                "title": "NestedJsonKeyRecord",
                "properties": {
                  "sub_id": {
                    "type": "integer"
                  }
                },
                "required": ["sub_id"]
              }
            },
            "required": ["idx", "nested"],
            "additionalProperties": false
        }))
        .unwrap()
    }

    fn payload_schema_string(&self) -> String {
        serde_json::to_string(&json!({
          "$schema": "http://json-schema.org/draft-07/schema#",
          "title": "JsonValue",
          "type": "object",
          "properties": {
            "value": {
              "type": "string"
            }
          },
          "required": ["value"],
          "additionalProperties": false
        }))
        .unwrap()
    }

    fn schema_type_string(&self) -> Option<String> {
        Some("JSON".to_string())
    }
}

struct JsonRawTestDataEncoder {}

impl JsonRawTestDataEncoder {
    fn new() -> Self {
        Self {}
    }
}

#[async_trait::async_trait]
impl TestDataEncoder for JsonRawTestDataEncoder {
    async fn key_for_idx<'a>(&'a self, idx: usize, _: &'a str) -> Vec<u8> {
        serde_json::to_vec(&json!({
            "key": idx,
        }))
        .unwrap()
    }

    async fn payload_for_idx<'a>(&'a self, idx: usize, _: &'a str) -> Vec<u8> {
        serde_json::to_vec(&json!({
            "payload": idx,
        }))
        .unwrap()
    }

    fn key_schema_string(&self) -> String {
        panic!("not implemented")
    }

    fn payload_schema_string(&self) -> String {
        panic!("not implemented")
    }

    fn schema_type_string(&self) -> Option<String> {
        None
    }
}
