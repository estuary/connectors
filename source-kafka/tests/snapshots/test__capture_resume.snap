---
source: tests/test.rs
expression: snap
---
["acmeCo/avro-topic",{"_meta":{"headers":{"header-key":"header-value-0"},"offset":0,"op":"u","partition":0,"timestamp":{"creationTime":"1970-01-02T00:00:00Z"},"topic":"avro-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"},"idx":0,"nested":{"sub_id":0},"value":"value-0"}]
["acmeCo/avro-topic",{"_meta":{"headers":{"header-key":"header-value-3"},"offset":1,"op":"u","partition":0,"timestamp":{"creationTime":"1970-01-05T00:00:00Z"},"topic":"avro-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"},"idx":3,"nested":{"sub_id":3},"value":"value-3"}]
["acmeCo/avro-topic",{"_meta":{"headers":{"header-key":"header-value-6"},"offset":2,"op":"u","partition":0,"timestamp":{"creationTime":"1970-01-08T00:00:00Z"},"topic":"avro-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"},"idx":6,"nested":{"sub_id":6},"value":"value-6"}]
["acmeCo/avro-topic",{"_meta":{"headers":{"header-key":"header-value-0"},"offset":3,"op":"d","partition":0,"timestamp":{"creationTime":"1970-01-02T00:00:00Z"},"topic":"avro-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"},"idx":0,"nested":{"sub_id":0}}]
["acmeCo/avro-topic",{"_meta":{"headers":{"header-key":"header-value-7"},"offset":2,"op":"u","partition":1,"timestamp":{"creationTime":"1970-01-09T00:00:00Z"},"topic":"avro-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"},"idx":7,"nested":{"sub_id":7},"value":"value-7"}]
["acmeCo/avro-topic",{"_meta":{"headers":{"header-key":"header-value-1"},"offset":3,"op":"d","partition":1,"timestamp":{"creationTime":"1970-01-03T00:00:00Z"},"topic":"avro-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"},"idx":1,"nested":{"sub_id":1}}]
["acmeCo/avro-topic",{"_meta":{"headers":{"header-key":"header-value-8"},"offset":2,"op":"u","partition":2,"timestamp":{"creationTime":"1970-01-10T00:00:00Z"},"topic":"avro-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"},"idx":8,"nested":{"sub_id":8},"value":"value-8"}]
["acmeCo/avro-topic",{"_meta":{"headers":{"header-key":"header-value-2"},"offset":3,"op":"d","partition":2,"timestamp":{"creationTime":"1970-01-04T00:00:00Z"},"topic":"avro-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"},"idx":2,"nested":{"sub_id":2}}]
["acmeCo/json-schema-topic",{"_meta":{"headers":{"header-key":"header-value-0"},"offset":0,"op":"u","partition":0,"timestamp":{"creationTime":"1970-01-02T00:00:00Z"},"topic":"json-schema-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"},"idx":0,"nested":{"sub_id":0},"value":"value-0"}]
["acmeCo/json-schema-topic",{"_meta":{"headers":{"header-key":"header-value-3"},"offset":1,"op":"u","partition":0,"timestamp":{"creationTime":"1970-01-05T00:00:00Z"},"topic":"json-schema-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"},"idx":3,"nested":{"sub_id":3},"value":"value-3"}]
["acmeCo/json-schema-topic",{"_meta":{"headers":{"header-key":"header-value-6"},"offset":2,"op":"u","partition":0,"timestamp":{"creationTime":"1970-01-08T00:00:00Z"},"topic":"json-schema-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"},"idx":6,"nested":{"sub_id":6},"value":"value-6"}]
["acmeCo/json-schema-topic",{"_meta":{"headers":{"header-key":"header-value-0"},"offset":3,"op":"d","partition":0,"timestamp":{"creationTime":"1970-01-02T00:00:00Z"},"topic":"json-schema-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"},"idx":0,"nested":{"sub_id":0}}]
["acmeCo/json-schema-topic",{"_meta":{"headers":{"header-key":"header-value-7"},"offset":2,"op":"u","partition":1,"timestamp":{"creationTime":"1970-01-09T00:00:00Z"},"topic":"json-schema-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"},"idx":7,"nested":{"sub_id":7},"value":"value-7"}]
["acmeCo/json-schema-topic",{"_meta":{"headers":{"header-key":"header-value-1"},"offset":3,"op":"d","partition":1,"timestamp":{"creationTime":"1970-01-03T00:00:00Z"},"topic":"json-schema-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"},"idx":1,"nested":{"sub_id":1}}]
["acmeCo/json-schema-topic",{"_meta":{"headers":{"header-key":"header-value-8"},"offset":2,"op":"u","partition":2,"timestamp":{"creationTime":"1970-01-10T00:00:00Z"},"topic":"json-schema-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"},"idx":8,"nested":{"sub_id":8},"value":"value-8"}]
["acmeCo/json-schema-topic",{"_meta":{"headers":{"header-key":"header-value-2"},"offset":3,"op":"d","partition":2,"timestamp":{"creationTime":"1970-01-04T00:00:00Z"},"topic":"json-schema-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"},"idx":2,"nested":{"sub_id":2}}]
["acmeCo/json-raw-topic",{"_key":"eyJrZXkiOjB9","_meta":{"headers":{"header-key":"header-value-0"},"offset":0,"op":"u","partition":0,"timestamp":{"creationTime":"1970-01-02T00:00:00Z"},"topic":"json-raw-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"},"payload":0}]
["acmeCo/json-raw-topic",{"_key":"eyJrZXkiOjN9","_meta":{"headers":{"header-key":"header-value-3"},"offset":1,"op":"u","partition":0,"timestamp":{"creationTime":"1970-01-05T00:00:00Z"},"topic":"json-raw-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"},"payload":3}]
["acmeCo/json-raw-topic",{"_key":"eyJrZXkiOjZ9","_meta":{"headers":{"header-key":"header-value-6"},"offset":2,"op":"u","partition":0,"timestamp":{"creationTime":"1970-01-08T00:00:00Z"},"topic":"json-raw-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"},"payload":6}]
["acmeCo/json-raw-topic",{"_key":"eyJrZXkiOjB9","_meta":{"headers":{"header-key":"header-value-0"},"offset":3,"op":"d","partition":0,"timestamp":{"creationTime":"1970-01-02T00:00:00Z"},"topic":"json-raw-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"}}]
["acmeCo/json-raw-topic",{"_key":"eyJrZXkiOjd9","_meta":{"headers":{"header-key":"header-value-7"},"offset":2,"op":"u","partition":1,"timestamp":{"creationTime":"1970-01-09T00:00:00Z"},"topic":"json-raw-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"},"payload":7}]
["acmeCo/json-raw-topic",{"_key":"eyJrZXkiOjF9","_meta":{"headers":{"header-key":"header-value-1"},"offset":3,"op":"d","partition":1,"timestamp":{"creationTime":"1970-01-03T00:00:00Z"},"topic":"json-raw-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"}}]
["acmeCo/json-raw-topic",{"_key":"eyJrZXkiOjh9","_meta":{"headers":{"header-key":"header-value-8"},"offset":2,"op":"u","partition":2,"timestamp":{"creationTime":"1970-01-10T00:00:00Z"},"topic":"json-raw-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"},"payload":8}]
["acmeCo/json-raw-topic",{"_key":"eyJrZXkiOjJ9","_meta":{"headers":{"header-key":"header-value-2"},"offset":3,"op":"d","partition":2,"timestamp":{"creationTime":"1970-01-04T00:00:00Z"},"topic":"json-raw-topic","uuid":"DocUUIDPlaceholder-329Bb50aa48EAa9ef"}}]
["connectorState",{"updated":{"bindingStateV1":{"avro-topic":{"partitions":{"0":3,"1":3,"2":3}},"json-raw-topic":{"partitions":{"0":3,"1":3,"2":3}},"json-schema-topic":{"partitions":{"0":3,"1":3,"2":3}}}}}]
